qiskit/algorithms/evolvers/evolution_problem.py

```diff
@@ -44,15 +45,124 @@ def __init__(
                 free parameter must be within the ``hamiltonian``.
             hamiltonian_value_dict: If the Hamiltonian contains free parameters, this
                 dictionary maps all these parameters to values.
+
+        Raises:
+            ValueError: If non-positive time of evolution is provided.
+            ValueError: If no ``initial_state`` is provided.
+            ValueError: If not all parameter values are provided.
         """
 
+        self.t_param = t_param
+        self.hamiltonian_value_dict = hamiltonian_value_dict
         self.hamiltonian = hamiltonian
         self.time = time
         self.initial_state = initial_state
         self.aux_operators = aux_operators
-        self.t_param = t_param
-        self.hamiltonian_value_dict = hamiltonian_value_dict
+        self.truncation_threshold = truncation_threshold
+
+    @property
+    def hamiltonian(self) -> OperatorBase:
+        """Returns a hamiltonian."""
+        return self._hamiltonian
+
+    @hamiltonian.setter
+    def hamiltonian(self, hamiltonian) -> None:
+        """
+        Sets a hamiltonian and validates it.
+
+        Raises:
+            ValueError: If no Hamiltonian is provided.
+            ValueError: If Hamiltonian parameters cannot be bound with data provided.
+        """
+        self._hamiltonian = None
+        if hamiltonian is None:
+            raise ValueError(
+                "No ``hamiltonian`` provided for the EvolutionProblem. It is required."
+            )
+        # TODO SparsePauliOp does not have .parameters because it is not allowed to be parametrized.
+        #  Can we handle this better than with an if?
+        if not isinstance(hamiltonian, SparsePauliOp):
+            self._check_parameters(hamiltonian, self.hamiltonian_value_dict, self.t_param)
+
+        self._hamiltonian = hamiltonian
+
+    @property
+    def time(self) -> float:
+        """Returns time."""
+        return self._time
+
+    @time.setter
+    def time(self, time) -> None:
+        """
+        Sets time and validates it.
+
+        Raises:
+            ValueError: If time is not positive.
+        """
+        self._time = None
+        if time <= 0:
+            raise ValueError(
+                f"Time of evolution provided is not positive, detected ``time={time}``."
+            )
+        self._time = time
+
+    @property
+    def initial_state(self) -> Union[StateFn, QuantumCircuit]:
+        """Returns an initial state."""
+        return self._initial_state
+
+    @initial_state.setter
+    def initial_state(self, initial_state) -> None:
+        """
+        Sets an initial state and validates it.
+
+        Raises:
+            ValueError: If no initial state is provided.
+        """
+        self._initial_state = None
+        if initial_state is None:
+            raise ValueError(
+                "No ``initial_state`` provided for the EvolutionProblem. It is required."
+            )
```

Comment:
If the initial state param has a typehint, if None is not permitted then it would not be given as Optional. In this case I would argue that the check here is not really needed. Also why would the internal state be set to something invalid ie. the prior self._initial_state = None. I would think the last line in this setter, along with typehint for the param same as in the contructor, would suffice.